src/mlpack/methods/pca/decomposition_policies/stochastic_method.hpp

```diff
@@ -0,0 +1,299 @@
+   * @param eigvec Matrix to put eigenvectors (loadings) into.
+   * @param rank Rank of the decomposition.
+   */
+  void Apply(const arma::mat& data,
+             const arma::mat& centeredData,
+             arma::mat& transformedData,
+             arma::vec& eigVal,
+             arma::mat& eigvec,
+             const size_t rank)
+  {
+    PCAFunction f(centeredData);
+
+    // Adjust the step size if necessary.
+    SetStepSize(optimizer);
+
+    eigvec = arma::randn(centeredData.n_rows, rank);
+    optimizer.Optimize(f, eigvec);
+
+    // Project the samples to the principals.
+    transformedData = arma::trans(eigvec) * centeredData;
+
+    // Approximate eigenvalues and eigenvectors using Rayleigh–Ritz method.
+    arma::mat u, v;
+    arma::svd_econ(u, eigVal, v, transformedData);
+
+    // Now we must square the singular values to get the eigenvalues.
+    // In addition we must divide by the number of points, because the
+    // covariance matrix is X * X' / (N - 1).
+    eigVal %= eigVal / (data.n_cols - 1);
+  }
+
+  //! Get the optimizer.
+  OptimizerType Optimizer() const { return optimizer; }
+  //! Modify the optimizer.
+  OptimizerType& Optimizer() { return optimizer; }
+
+ private:
+  //! Adjust the step size if required.
+  template<typename T>
+  inline typename std::enable_if<
+      HasStepSize<T, double&(T::*)()>::value, void>::type
+  SetStepSize(T& optimizer)
+  {
+    if (optimizer.StepSize() > 0)
+    {
+      Log::Warn << "StepSize(): invalid value (> 0), automatically take the "
+          << "negative direction." << std::endl;
+      optimizer.StepSize() *= -1;
+    }
+  }
+
+  //! Adjust the step size if required.
+  template<typename T>
+  inline typename std::enable_if<
+      !HasStepSize<T, double&(T::*)()>::value, void>::type
+  SetStepSize(T& /* optimizer */)
+  {
+    /* Nothing to do here */
+  }
+
+  OptimizerType optimizer;
+
+  /**
+   * The principal component analysis (PCA) function for the
+   * Principal component analysis (PCA). This is used by various mlpack
+   * optimizers to be used for principal component analysis.
+   */
+  class PCAFunction
+  {
+   public:
+    /**
+     * Construct The principal component analysis (PCA) function with the
+     * given data.
+     *
+     * @param data Data matrix.
+     */
+    PCAFunction(const arma::mat& data) :
+        // We promise to be well-behaved... the elements won't be modified.
+        data(math::MakeAlias(const_cast<arma::mat&>(data), false)),
+        pl(0)
+    {
+      // Nothing to do here.
+    }
+
+    /**
+     * Shuffle the order of function visitation. This may be called by the
+     * optimizer.
+     */
+    void Shuffle()
+    {
+      // Generate ordering.
+      arma::uvec ordering = arma::shuffle(arma::linspace<arma::uvec>(0,
+          data.n_cols - 1, data.n_cols));
+      arma::mat newData = data.cols(ordering);
+
+      // If we are an alias, make sure we don't write to the original data.
+      math::ClearAlias(data);
```

Comment:
Hm, I think that even if `data` is an alias, when we call `data = std::move(newData)` then `data` will no longer be an alias since it just takes the memory pointer from `newData` (which is not an alias).  I am not 100% sure but about 95% sure on that, based on looking at the Armadillo matrix `operator=` for rvalue references.

Reply:
Right, at this point I copied:

https://github.com/mlpack/mlpack/blob/e3fe135459593bf5276b3a2794513a9de2ec1c51/src/mlpack/methods/logistic_regression/logistic_regression_function_impl.hpp#L68-L80

Reply:
(Sorry for the slow response here.)  I see, I guess I originally wrote this code.  I think that maybe we could omit the `ClearAlias()` call in both places, but I suppose it's not a huge deal either way---up to you if you want to make the change.